ssl/record/rec_layer_s3.c

```diff
@@ -1583,14 +1549,53 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
         goto start;
     }
 
-    if (s->shutdown & SSL_SENT_SHUTDOWN) { /* but we have not received a
-                                            * shutdown */
+    /*
+     * If we've sent a close_notify but not yet received one back then ditch
+     * anything we read.
+     */
+    if ((s->shutdown & SSL_SENT_SHUTDOWN) != 0 ) {
+        /*
+         * In TLSv1.3 this could get problematic if we receive a KeyUpdate
+         * message after we sent a close_notify because we're about to ditch it,
+         * so we won't be able to read a close_notify sent afterwards! We don't
+         * support that.
+         */
         s->rwstate = SSL_NOTHING;
         SSL3_RECORD_set_length(rr, 0);
         SSL3_RECORD_set_read(rr);
+
+        if (SSL3_RECORD_get_type(rr) == SSL3_RT_HANDSHAKE
+                && (s->mode & SSL_MODE_AUTO_RETRY) != 0)
+            goto start;
         return 0;
```

Comment:
If I understand correctly, if auto retry is off you get a fatal error.

Reply:
It also seems trivial to add a call to SSLfatal() in case the type was application data.

Reply:
If this is a handshake message then it should probably be processed, instead of discarded?
I mean something like that (or maybe more specific which message exactly are expected):
```
diff --git a/ssl/record/rec_layer_s3.c b/ssl/record/rec_layer_s3.c
index 61010f4..0b983df 100644
--- a/ssl/record/rec_layer_s3.c
+++ b/ssl/record/rec_layer_s3.c
@@ -1583,14 +1583,6 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
         goto start;
     }
 
-    if (s->shutdown & SSL_SENT_SHUTDOWN) { /* but we have not received a
-                                            * shutdown */
-        s->rwstate = SSL_NOTHING;
-        SSL3_RECORD_set_length(rr, 0);
-        SSL3_RECORD_set_read(rr);
-        return 0;
-    }
-
     if (SSL3_RECORD_get_type(rr) == SSL3_RT_CHANGE_CIPHER_SPEC) {
         SSLfatal(s, SSL_AD_UNEXPECTED_MESSAGE, SSL_F_SSL3_READ_BYTES,
                  SSL_R_CCS_RECEIVED_EARLY);
@@ -1644,6 +1636,14 @@ int ssl3_read_bytes(SSL *s, int type, int *recvd_type, unsigned char *buf,
         goto start;
     }
 
+    if (s->shutdown & SSL_SENT_SHUTDOWN) { /* but we have not received a
+                                            * shutdown */
+        s->rwstate = SSL_NOTHING;
+        SSL3_RECORD_set_length(rr, 0);
+        SSL3_RECORD_set_read(rr);
+        return 0;
+    }
+
     switch (SSL3_RECORD_get_type(rr)) {
     default:
         /*
```

Reply:
> I think you want to make sure to return SSL_ERROR_WANT_READ in case auto retry is off and the type was handshake data, like it's done below. That probably also means returning -1 for compatibility reasons.

Yes, this makes sense. I made that change.

> It also seems trivial to add a call to SSLfatal() in case the type was application data.

I think this could be a breaking change if we did that. At the moment you can just keep calling SSL_shutdown(), ignoring any errors, until you eventually get a 1 return. If you call SSLfatal() it marks the SSL object as unusable, and any subsequent calls to SSL_shutdown() will immediately fail, i.e. you will never get a 1 return. We *could* just call SSLerr - but that might risk lots of stale errors on the queue.

Reply:
> If this is a handshake message then it should probably be processed, instead of discarded?

Hmmm...that's non-trivial and throws up various issues if we attempt that. In TLSv1.2 and below I think it is always the right thing to do to discard any handshake messages. You cannot shutdown while still in a handshake, so we are in a post-handshake state if we have already sent close_notify. The only post-handshake messages in TLSv1.2 are reneg related. Since we have already shutdown the "write" side of the connection we cannot possibly complete a reneg. In TLSv1.3 much the same argument applies for post-handshake auth messages. We *could* choose to process NewSessionTickets but it does no harm to ignore them and it simplifies things considerably if we do. KeyUpdates are a bit tricky. The comment in the code basically says we ignore that problem. If we choose to process them, then effectively we have to pass all handshake messages to the state machine (we don't want the record layer making state machine decisions - that gets us back to where we were before the state machine refactor). We'd then have to make sure we correctly filter out in the state machine all handshake messages that we don't want to handle if we are in a half-shutdown state and only process KeyUpdate. We'd also not be able to respond with a KeyUpdate of our own where the peer has requested that - so there would have to be some special case code for that. In my mind the KeyUpdate problem is a corner case that is going to be rare in practice - so lets not complicate everything by trying to support it. The simplest thing to do is just ignore all handshake messages.

Reply:
Hmmm....that sounds like a 1.2.0 thing.

Reply:
I have no idea why you think that's a 1.2.0 thing. I think it's a bug fix.

Reply:
Because we previously we considered that receiving application data during shutdown was a non fatal, recoverable, error. You want to change it into a fatal error. That's a behaviour change not a bug fix.

Reply:
My recollection is that SSL_shutdown returns a SSL_ERROR_SSL.
And if an application is retrying SSL_shutdown it is unpredictable if
the SSL will recover or crash.

Reply:
It actually returns SSL_ERROR_SYSCALL, because nobody bothered to actually say that there was an error, so we just return SSL_ERROR_SYSCALL. But both are fatal errors. First returning a fatal error and then returning it was successful after all doesn't make sense. Calling SSLfatal() should turn it into a SSL_ERROR_SSL.

Reply:
Hmmm....I suppose existing application code has to be able to handle fatal errors to deal with stuff like the underlying TCP connection closing without a close_notify...so this would appear the same as if that occurred.

Reply:
I made this case a fatal error as requested.